src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,360 @@
+             G.mem_incidence_sets_iff_eq H_adj, mul_one],
+  rw adj_iff_exists_edge_coe at H_adj,
+  rcases H_adj with ⟨e, H_e⟩,
+  simp only [←H_e, subtype.ext_iff.symm, filter_eq', filter_congr_decidable,
+             if_true, mem_univ, nat.cast_one, card_singleton]
+end
+
+theorem inc_matrix_mul_non_adj {i j : V} {e : sym2 V} (Hne : i ≠ j) (H_non_adj : ¬G.adj i j)
+[char_zero R] :
+  G.inc_matrix R i e * G.inc_matrix R j e = 0 :=
+begin
+  by_cases H₁ : G.inc_matrix R i e = 0,
+  { rw [H₁, zero_mul] },
+  { rw [inc_matrix_ne_zero, inc_matrix_coeff_eq_one_iff] at H₁,
+    by_cases H₂ : G.inc_matrix R j e = 0,
+    { rw [H₂, mul_zero] },
+    { rw [inc_matrix_ne_zero, inc_matrix_coeff_eq_one_iff] at H₂,
+      exfalso,
+      apply H_non_adj,
+      rw [←mem_edge_set, ←G.edge_in_two_incidence_sets Hne ⟨H₁, H₂⟩],
+      exact set.mem_of_mem_inter_left H₁ } }
+end
+
+/-- Every element from `M` is idempotent. -/
+theorem inc_matrix_element_power_id {i : V} {e : sym2 V} :
+  (G.inc_matrix R i e) * (G.inc_matrix R i e) = G.inc_matrix R i e :=
+by simp [inc_matrix_apply]
+
+/-- The sum of elements from row `i` of `M` is equal to the `degree` of vertex `i`. -/
+theorem degree_equals_sum_of_incidence_row {i : V} [char_zero R] :
+  (G.degree i : R) = ∑ e, G.inc_matrix R i e :=
+begin
+  simp only [inc_matrix_apply, sum_boole],
+  rw [←card_incidence_set_eq_degree, nat.cast_inj],
+  refine finset.card_congr _ _ _ _,
+  { rintro ⟨e, he⟩ he',
+    exact e },
+  { rintro ⟨e, he⟩ he',
+    simpa only [true_and, finset.mem_univ, finset.mem_filter] using he },
+  { rintro ⟨e1, he1⟩ ⟨e2, he2⟩ he1' he2' hr,
+    ext,
+    simp only [subtype.mk_eq_mk] at hr,
+    simp only [hr] },
+  { rintro e he,
+    use e,
+    { simpa only [true_and, finset.mem_univ, finset.mem_filter] using he },
+    { simp only [finset.mem_univ, exists_prop_of_true] } }
+end
+
+end incidence
+
+/-! ## Orientations -/
+
+section orientations
+
+/-- Define an `orientation` on the simple graph `G` as a structure that defines (consistently)
+for each edge a `head` and a `tail`. -/
+@[ext]
+structure orientation (G : simple_graph V) :=
+(head : G.edge_set → V)
+(tail : G.edge_set → V)
+(consistent (e : G.edge_set) : ↑e = ⟦(head(e), tail(e))⟧)
+
+noncomputable instance (G : simple_graph V) : inhabited (orientation G) :=
+⟨{ head := λ (e : G.edge_set), (quotient.out (e : sym2 V)).fst,
+  tail := λ (e : G.edge_set), (quotient.out (e : sym2 V)).snd,
+  consistent := λ (e : G.edge_set), by rw [prod.mk.eta, quotient.out_eq] }⟩
+
+variables {o : orientation G}
+
+lemma head_tail (o : orientation G) (i : V) (e : G.edge_set) :
+  i = o.head e ∨ i = o.tail e ∨ (i ≠ o.head e ∧ i ≠ o.tail e) :=
+by tauto
+
+lemma edge_not_incident {i j : V} {e : G.edge_set}
+  (H_e : ¬↑e = ⟦(i, j)⟧) (H_adj : G.adj i j) : ↑e ∉ G.incidence_set i ∨ ↑e ∉ G.incidence_set j :=
+begin
+  classical,
+  by_contradiction h,
+  rw [decidable.not_or_iff_and_not, not_not, not_not] at h,
```

Comment:
No point in using the `decidable` lemma once you went `classical`! Also, this lemma really is the (classical) contrapositive of `edge_in_two_incidence_sets`, so I would encourage you to nuke it and rather perform this `¬` distribution directly in the proof of `oriented_inc_matrix_mul_of_adj`.